File: aicq-server/src/services/accountService.ts
import { v4 as uuidv4 } from 'uuid';
import crypto from 'crypto';
import bcrypt from 'bcrypt';
import nacl from 'tweetnacl';
import { store } from '../db/memoryStore';
import { Account, Session, AccountType, FriendPermission } from '../models/types';

const BCRYPT_ROUNDS = 12;

// ─── Password Hashing (bcrypt with SHA-256 legacy fallback) ────────

async function hashPassword(password: string): Promise<string> {
  return bcrypt.hash(password, BCRYPT_ROUNDS);
}

async function verifyPassword(password: string, storedHash: string): Promise<boolean> {
  if (storedHash.startsWith('$2b$') || storedHash.startsWith('$2a$')) {
    return bcrypt.compare(password, storedHash);
  }
  // Legacy SHA-256 fallback
  const [salt, hash] = storedHash.split(':');
  const computed = crypto.createHash('sha256').update(password + salt).digest('hex');
  return computed === hash;
}

// ─── JWT Token ────────────────────────────────────────────────────

// Generate JWT token (simple implementation, replace with jsonwebtoken in production)
function generateToken(payload: Record<string, unknown>, secret: string, expiresIn: number): { token: string; expiresAt: number } {
  const header = Buffer.from(JSON.stringify({ alg: 'HS256', typ: 'JWT' })).toString('base64url');
  const now = Math.floor(Date.now() / 1000);

  const body = Buffer.from(JSON.stringify({
    ...payload,
    iat: now,
    exp: now + expiresIn,
  })).toString('base64url');

  const signature = crypto
    .createHmac('sha256', secret)
    .update(`${header}.${body}`)
    .digest('base64url');

  const token = `${header}.${body}.${signature}`;
  return { token, expiresAt: now + expiresIn };
}

/**
 * Verify a JWT token. Checks signature, algorithm, and expiration.
 * Exported for use by auth middleware.
 */
export function verifyJWT(token: string, secret: string): Record<string, unknown> | null {
  try {
    const parts = token.split('.');
    if (parts.length !== 3) return null;

    const header = Buffer.from(parts[0], 'base64url').toString();
    const headerObj = JSON.parse(header);
    if (headerObj.alg !== 'HS256') return null;

    // Verify the signature
    const expectedSig = crypto
      .createHmac('sha256', secret)
      .update(`${parts[0]}.${parts[1]}`)
      .digest('base64url');

    if (expectedSig !== parts[2]) return null;

    const payload = JSON.parse(Buffer.from(parts[1], 'base64url').toString());

    // Check expiration
    const now = Math.floor(Date.now() / 1000);
    if (payload.exp && (payload.exp as number) < now) return null;

    return payload;
  } catch {
    return null;
  }
}

// ─── Public API ────────────────────────────────────────────────────

export async function registerHuman(
  target: string,
  type: 'email' | 'phone',
  code: string,
  password: string,
  displayName: string,
  publicKey: string
): Promise<Account> {
  // Verify the code
  const key = `${type}:${target}:register`;
  const codeRecord = store.verificationCodes.get(key);
  if (!codeRecord || codeRecord.code !== code) {
    throw new Error('验证码错误');
  }
  if (codeRecord.expiresAt <= Date.now()) {
    store.verificationCodes.delete(key);
    throw new Error('验证码已过期');
  }
  store.verificationCodes.delete(key);

  // Check if already registered
  const existing = findAccountByTarget(target, type);
  if (existing) {
    throw new Error('该' + (type === 'email' ? '邮箱' : '手机号') + '已注册');
  }

  const accountId = uuidv4();
  const passwordHash = await hashPassword(password);

  const account: Account = {
    id: accountId,
    type: 'human',
    email: type === 'email' ? target : undefined,
    phone: type === 'phone' ? target : undefined,
    passwordHash,
    displayName: displayName || (type === 'email' ? target.split('@')[0] : target),
    publicKey,
    createdAt: Date.now(),
    lastLoginAt: Date.now(),
    status: 'active',
    friends: [],
    maxFriends: 200,
    friendPermissions: {},
    visitPermissions: [],
  };

  store.setAccount(account);
  console.log(`[account] Human registered: ${target} (${accountId})`);
  return account;
}

export async function loginHuman(
  target: string,
  type: 'email' | 'phone',
  password?: string,
  code?: string
): Promise<{ account: Account; session: Session }> {
  const account = findAccountByTarget(target, type);
  if (!account) {
    throw new Error('账号不存在');
  }
  if (account.status !== 'active') {
    throw new Error('账号已被禁用');
  }

  // Email login: verify password
  if (type === 'email' && password) {
    if (!account.passwordHash) {
      throw new Error('该账号未设置密码');
    }
    const valid = await verifyPassword(password, account.passwordHash);
    if (!valid) {
      throw new Error('密码错误');
    }
  }

  // Phone login: verify code
  if (type === 'phone' && code) {
    const key = `phone:${target}:login`;
    const codeRecord = store.verificationCodes.get(key);
    if (!codeRecord || codeRecord.code !== code) {
      throw new Error('验证码错误');
    }
    if (codeRecord.expiresAt <= Date.now()) {
      store.verificationCodes.delete(key);
      throw new Error('验证码已过期');
    }
    store.verificationCodes.delete(key);
  }

  if (!password && !code) {
    throw new Error('请提供密码或验证码');
  }

  // Update last login
  account.lastLoginAt = Date.now();
  store.setAccount(account);

  // Create session
  const session = createSession(account);
  return { account, session };
}

// ─── AI Agent Auth ─────────────────────────────────────────────

// Active challenges for agent auth
const agentChallenges = new Map<string, { challenge: string; publicKey: string; agentName?: string; expiresAt: number }>();

export function requestAgentChallenge(publicKey: string, agentName?: string): { challenge: string; challengeId: string; expiresAt: number } {
  const challengeId = uuidv4();
  const challenge = uuidv4();

  agentChallenges.set(challengeId, {
    challenge,
    publicKey,
    agentName,
    expiresAt: Date.now() + 5 * 60 * 1000, // 5 min
  });

  return { challenge, challengeId, expiresAt: Date.now() + 5 * 60 * 1000 };
}

export async function loginAgent(
  publicKey: string,
  agentName: string | undefined,
  signature: string,
  challengeId: string
): Promise<{ account: Account; session: Session }> {
  const challengeData = agentChallenges.get(challengeId);
  if (!challengeData) {
    throw new Error('挑战不存在或已过期');
  }
  if (Date.now() > challengeData.expiresAt) {
    agentChallenges.delete(challengeId);
    throw new Error('挑战已过期，请重新发起');
  }
  if (challengeData.publicKey !== publicKey) {
    throw new Error('公钥与挑战不匹配');
  }

  agentChallenges.delete(challengeId);

  // Verify signature using tweetnacl
  const isValid = verifyAgentSignature(publicKey, challengeData.challenge, signature);
  if (!isValid) {
    throw new Error('签名验证失败');
  }

  // Find or auto-create account
  let account = findAccountByPublicKey(publicKey);
  if (!account) {
    account = createAgentAccount(publicKey, agentName || 'AI Agent');
  }

  account.lastLoginAt = Date.now();
  store.setAccount(account);

  const session = createSession(account);
  return { account, session };
}

function verifyAgentSignature(publicKeyBase64: string, challenge: string, signatureBase64: string): boolean {
  try {
    const message = Buffer.from(challenge, 'utf8');
    const signature = Buffer.from(signatureBase64, 'base64');
    const publicKey = Buffer.from(publicKeyBase64, 'base64');

    if (publicKey.length !== 32 || signature.length !== 64) return false;

    return nacl.sign.detached.verify(message, signature, publicKey);
  } catch {
    return false;
  }
}

function createAgentAccount(publicKey: string, agentName: string): Account {
  const accountId = uuidv4();
  const account: Account = {
    id: accountId,
    type: 'ai',
    agentName,
    publicKey,
    createdAt: Date.now(),
    lastLoginAt: Date.now(),
    status: 'active',
    friends: [],
    maxFriends: 200,
    friendPermissions: {},
    visitPermissions: [],
  };
  store.setAccount(account);
  console.log(`[account] AI Agent auto-registered: ${agentName} (${accountId})`);
  return account;
}

// ─── Helpers ────────────────────────────────────────────────────

/** O(1) lookup by email or phone using index maps */
function findAccountByTarget(target: string, type: 'email' | 'phone'): Account | undefined {
  const index = type === 'email' ? store.emailIndex : store.phoneIndex;
  const accountId = index.get(target);
  return accountId ? store.accounts.get(accountId) : undefined;
}

/** O(1) lookup by publicKey using index map */
function findAccountByPublicKey(publicKey: string): Account | undefined {
  const accountId = store.publicKeyIndex.get(publicKey);
  return accountId ? store.accounts.get(accountId) : undefined;
}

export function createSession(account: Account): Session {
  const jwtSecret = process.env.JWT_SECRET || 'aicq-default-jwt-secret-change-in-production';
  const { token, expiresAt } = generateToken(
    { sub: account.id, type: account.type, displayName: account.displayName || account.agentName },
    jwtSecret,
    3600 // 1 hour
  );
  const { token: refreshToken, expiresAt: refreshExpiresAt } = generateToken(
    { sub: account.id },
    jwtSecret + '-refresh',
    30 * 24 * 3600 // 30 days
  );

  const session: Session = {
    id: uuidv4(),
    accountId: account.id,
    token,
    refreshToken,
    createdAt: Date.now(),
    expiresAt: expiresAt * 1000,
  };

  store.sessions.set(session.id, session);
  return session;
}

export function refreshSession(refreshToken: string): Session | null {
  const jwtSecret = process.env.JWT_SECRET || 'aicq-default-jwt-secret-change-in-production';
  const payload = verifyJWT(refreshToken, jwtSecret + '-refresh');
  if (!payload) return null;

  const now = Math.floor(Date.now() / 1000);
  if ((payload.exp as number) < now) return null;

  const account = store.accounts.get(payload.sub as string);
  if (!account || account.status !== 'active') return null;

  return createSession(account);
}

export function getAccount(accountId: string): Account | undefined {
  return store.accounts.get(accountId);
}

export function cleanupExpiredSessions(): number {
  const now = Date.now();
  let removed = 0;
  for (const [sessionId, session] of store.sessions) {
    if (session.expiresAt <= now) {
      store.sessions.delete(sessionId);
      removed++;
    }
  }
  return removed;
}

export function cleanupExpiredChallenges(): number {
  const now = Date.now();
  let removed = 0;
  for (const [id, data] of agentChallenges) {
    if (data.expiresAt <= now) {
      agentChallenges.delete(id);
      removed++;
    }
  }
  return removed;
}

File: server/services/account_service.py
"""
AICQ Account Service
=====================
Handles account registration, authentication, JWT token management,
and Ed25519 challenge-response authentication for AI agents.
"""

from __future__ import annotations

import logging
import uuid
from datetime import datetime, timezone, timedelta
from typing import Any, Dict, Optional

import bcrypt
import jwt
from nacl.signing import VerifyKey
from nacl.exceptions import BadSignatureError

from db import DatabaseManager, json_serialize, json_list, iso_now
from config import Config

logger = logging.getLogger("aicq.account_service")


# ─── Password Helpers ───────────────────────────────────────────────────


def _hash_password(password: str) -> str:
    """Hash a plaintext password using bcrypt. Returns the UTF-8 decoded hash."""
    salt = bcrypt.gensalt(rounds=12)
    hashed = bcrypt.hashpw(password.encode("utf-8"), salt)
    return hashed.decode("utf-8")


def _verify_password(password: str, password_hash: str) -> bool:
    """Verify a plaintext password against a bcrypt hash."""
    try:
        return bcrypt.checkpw(password.encode("utf-8"), password_hash.encode("utf-8"))
    except Exception:
        return False


# ─── JWT Helpers ────────────────────────────────────────────────────────


def _create_access_token(account_id: str, account_type: str, display_name: Optional[str]) -> str:
    """Create a short-lived access token (1 hour by default)."""
    now = datetime.now(timezone.utc)
    expiry = now + timedelta(seconds=Config.JWT_ACCESS_TOKEN_EXPIRY)
    payload = {
        "sub": account_id,
        "type": account_type,
        "displayName": display_name or "",
        "iat": now,
        "exp": expiry,
    }
    return jwt.encode(payload, Config.JWT_SECRET, algorithm="HS256")


def _create_refresh_token(account_id: str) -> str:
    """Create a long-lived refresh token (30 days by default)."""
    now = datetime.now(timezone.utc)
    expiry = now + timedelta(seconds=Config.JWT_REFRESH_TOKEN_EXPIRY)
    payload = {
        "sub": account_id,
        "type": "refresh",
        "iat": now,
        "exp": expiry,
    }
    return jwt.encode(payload, Config.JWT_SECRET + "-refresh", algorithm="HS256")


# ─── Registration ──────────────────────────────────────────────────────


async def register_human(
    db: DatabaseManager,
    email: str,
    phone: Optional[str],
    password: str,
    display_name: str,
    public_key: str,
) -> Dict[str, Any]:
    """Register a new human account.

    Returns the created account record (without password_hash).
    Raises ValueError if the email or public_key is already registered.
    """
    # Check for duplicate email
    if email:
        existing = await db.fetchone("SELECT id FROM accounts WHERE email = ?", (email,))
        if existing:
            raise ValueError("Email already registered")

    # Check for duplicate public_key
    existing = await db.fetchone("SELECT id FROM accounts WHERE public_key = ?", (public_key,))
    if existing:
        raise ValueError("Public key already registered")

    account_id = uuid.uuid4().hex
    now = iso_now()
    password_hash = _hash_password(password)

    await db.execute(
        """
        INSERT INTO accounts
            (id, type, email, phone, password_hash, display_name, public_key,
             created_at, status, friends, visit_permissions)
        VALUES (?, 'human', ?, ?, ?, ?, ?, ?, 'active', '[]', '[]')
        """,
        (account_id, email, phone, password_hash, display_name, public_key, now),
    )

    account = await get_account(db, account_id)
    logger.info("Registered human account: %s (%s)", account_id, email)
    return account


async def register_ai_agent(
    db: DatabaseManager,
    public_key: str,
    agent_name: str,
) -> Dict[str, Any]:
    """Register a new AI agent account.

    Returns the created account record.
    Raises ValueError if the public_key is already registered.
    """
    existing = await db.fetchone("SELECT id FROM accounts WHERE public_key = ?", (public_key,))
    if existing:
        raise ValueError("Public key already registered")

    account_id = uuid.uuid4().hex
    now = iso_now()

    await db.execute(
        """
        INSERT INTO accounts
            (id, type, agent_name, public_key, created_at, status, friends, visit_permissions)
        VALUES (?, 'ai', ?, ?, ?, 'active', '[]', '[]')
        """,
        (account_id, agent_name, public_key, now),
    )

    account = await get_account(db, account_id)
    logger.info("Registered AI agent account: %s (%s)", account_id, agent_name)
    return account


# ─── Login / Authentication ────────────────────────────────────────────


async def login_human(
    db: DatabaseManager,
    email: str,
    password: str,
    device_info: Optional[str] = None,
) -> Dict[str, Any]:
    """Authenticate a human account by email + password.

    Returns a dict with access_token, refresh_token, and account info.
    Raises ValueError on invalid credentials.
    """
    account = await db.fetchone("SELECT * FROM accounts WHERE email = ? AND type = 'human'", (email,))
    if not account:
        raise ValueError("Invalid email or password")

    if account["status"] != "active":
        raise ValueError("Account is not active")

    if not account["password_hash"] or not _verify_password(password, account["password_hash"]):
        raise ValueError("Invalid email or password")

    # Update last login
    now = iso_now()
    await db.execute("UPDATE accounts SET last_login_at = ? WHERE id = ?", (now, account["id"]))

    # Create session
    tokens = await create_session(db, account["id"], device_info)

    logger.info("Human login: %s", account["id"])
    return {
        "access_token": tokens["access_token"],
        "refresh_token": tokens["refresh_token"],
        "account": _sanitize_account(account),
    }


async def login_human_phone(
    db: DatabaseManager,
    phone: str,
    code: str,
    device_info: Optional[str] = None,
) -> Dict[str, Any]:
    """Authenticate a human account by phone + verification code.

    Verifies the code, then creates a session.
    Raises ValueError on invalid phone, code, or account state.
    """
    account = await db.fetchone("SELECT * FROM accounts WHERE phone = ? AND type = 'human'", (phone,))
    if not account:
        raise ValueError("No account found for this phone number")

    if account["status"] != "active":
        raise ValueError("Account is not active")

    # Verify the code
    from services.verification_service import verify_code

    verified = await verify_code(db, phone, code, "login")
    if not verified:
        raise ValueError("Invalid or expired verification code")

    # Update last login
    now = iso_now()
    await db.execute("UPDATE accounts SET last_login_at = ? WHERE id = ?", (now, account["id"]))

    # Create session
    tokens = await create_session(db, account["id"], device_info)

    logger.info("Human phone login: %s", account["id"])
    return {
        "access_token": tokens["access_token"],
        "refresh_token": tokens["refresh_token"],
        "account": _sanitize_account(account),
    }


async def challenge_agent(
    db: DatabaseManager,
    public_key: str,
) -> Dict[str, Any]:
    """Generate and store a random challenge for Ed25519 agent authentication.

    Returns a dict with session_id and challenge (hex-encoded).
    Raises ValueError if no AI account matches the public_key.
    """
    account = await db.fetchone(
        "SELECT id FROM accounts WHERE public_key = ? AND type = 'ai'",
        (public_key,),
    )
    if not account:
        raise ValueError("No AI agent account found for this public key")

    # Generate a random challenge (32 bytes, hex-encoded)
    import secrets as _secrets
    challenge = _secrets.token_hex(32)

    # Store challenge in a lightweight way — we use the sessions table
    # with a special prefix on the token field to distinguish challenge sessions
    session_id = uuid.uuid4().hex
    now = iso_now()
    # Challenge expires in 5 minutes
    challenge_expiry = (datetime.now(timezone.utc) + timedelta(minutes=5)).isoformat()

    await db.execute(
        """
        INSERT INTO sessions (id, account_id, token, refresh_token, device_info, created_at, expires_at)
        VALUES (?, ?, ?, '', 'challenge', ?, ?)
        """,
        (session_id, account["id"], f"challenge:{challenge}", now, challenge_expiry),
    )

    return {"session_id": session_id, "challenge": challenge}


async def login_agent(
    db: DatabaseManager,
    public_key: str,
    signature: str,
    challenge: str,
    device_info: Optional[str] = None,
) -> Dict[str, Any]:
    """Authenticate an AI agent via Ed25519 challenge-response.

    The agent signs the challenge with its private key; we verify using
    the stored public_key. On success, create a session and return tokens.

    Raises ValueError on invalid credentials or expired challenge.
    """
    account = await db.fetchone(
        "SELECT * FROM accounts WHERE public_key = ? AND type = 'ai'",
        (public_key,),
    )
    if not account:
        raise ValueError("No AI agent account found for this public key")

    if account["status"] != "active":
        raise ValueError("Account is not active")

    # Verify the Ed25519 signature
    try:
        # public_key and signature are hex-encoded
        verify_key = VerifyKey(bytes.fromhex(public_key))
        verify_key.verify(bytes.fromhex(challenge), bytes.fromhex(signature))
    except (BadSignatureError, Exception) as exc:
        raise ValueError(f"Invalid signature: {exc}") from exc

    # Clean up challenge sessions
    await db.execute(
        "DELETE FROM sessions WHERE account_id = ? AND device_info = 'challenge'",
        (account["id"],),
    )

    # Update last login
    now = iso_now()
    await db.execute("UPDATE accounts SET last_login_at = ? WHERE id = ?", (now, account["id"]))

    # Create session
    tokens = await create_session(db, account["id"], device_info)

    logger.info("AI agent login: %s", account["id"])
    return {
        "access_token": tokens["access_token"],
        "refresh_token": tokens["refresh_token"],
        "account": _sanitize_account(account),
    }


# ─── Session / Token Management ────────────────────────────────────────


async def create_session(
    db: DatabaseManager,
    account_id: str,
    device_info: Optional[str] = None,
) -> Dict[str, str]:
    """Create a new session with access and refresh tokens.

    Returns a dict with access_token and refresh_token.
    """
    account = await db.fetchone("SELECT id, type, display_name, agent_name FROM accounts WHERE id = ?", (account_id,))
    if not account:
        raise ValueError("Account not found")

    display_name = account["display_name"] or account["agent_name"] or ""
    access_token = _create_access_token(account_id, account["type"], display_name)
    refresh_token = _create_refresh_token(account_id)

    session_id = uuid.uuid4().hex
    now = iso_now()
    refresh_expiry = (datetime.now(timezone.utc) + timedelta(seconds=Config.JWT_REFRESH_TOKEN_EXPIRY)).isoformat()

    await db.execute(
        """
        INSERT INTO sessions (id, account_id, token, refresh_token, device_info, created_at, expires_at)
        VALUES (?, ?, ?, ?, ?, ?, ?)
        """,
        (session_id, account_id, access_token, refresh_token, device_info, now, refresh_expiry),
    )

    return {"access_token": access_token, "refresh_token": refresh_token}


def verify_token(token: str) -> Dict[str, Any]:
    """Decode and verify a JWT access token.

    Returns the token payload dict.
    Raises jwt.PyJWTError on invalid or expired tokens.
    """
    return jwt.decode(token, Config.JWT_SECRET, algorithms=["HS256"])


async def refresh_token(db: DatabaseManager, refresh_tok: str) -> Dict[str, Any]:
    """Verify a refresh token and issue a new access token.

    Returns a dict with new access_token and refresh_token.
    Raises ValueError on invalid or expired refresh tokens.
    """
    try:
        payload = jwt.decode(refresh_tok, Config.JWT_SECRET + "-refresh", algorithms=["HS256"])
    except jwt.PyJWTError as exc:
        raise ValueError(f"Invalid refresh token: {exc}") from exc

    account_id = payload.get("sub")
    if not account_id:
        raise ValueError("Invalid refresh token: missing subject")

    # Verify the account still exists and is active
    account = await db.fetchone(
        "SELECT id, type, status, display_name, agent_name FROM accounts WHERE id = ?",
        (account_id,),
    )
    if not account:
        raise ValueError("Account not found")
    if account["status"] != "active":
        raise ValueError("Account is not active")

    # Verify the session still exists
    session = await db.fetchone(
        "SELECT id FROM sessions WHERE refresh_token = ? AND account_id = ?",
        (refresh_tok, account_id),
    )
    if not session:
        raise ValueError("Session not found")

    # Delete the old session
    await db.execute("DELETE FROM sessions WHERE id = ?", (session["id"],))

    # Create a new session
    tokens = await create_session(db, account_id)

    logger.info("Refreshed token for account: %s", account_id)
    return tokens


# ─── Account CRUD ──────────────────────────────────────────────────────


async def get_account(db: DatabaseManager, account_id: str) -> Optional[Dict[str, Any]]:
    """Fetch an account by ID. Returns sanitized dict or None."""
    account = await db.fetchone("SELECT * FROM accounts WHERE id = ?", (account_id,))
    if account is None:
        return None
    return _sanitize_account(account)


async def get_account_by_email(db: DatabaseManager, email: str) -> Optional[Dict[str, Any]]:
    """Fetch an account by email. Returns sanitized dict or None."""
    account = await db.fetchone("SELECT * FROM accounts WHERE email = ?", (email,))
    if account is None:
        return None
    return _sanitize_account(account)


async def get_account_by_phone(db: DatabaseManager, phone: str) -> Optional[Dict[str, Any]]:
    """Fetch an account by phone. Returns sanitized dict or None."""
    account = await db.fetchone("SELECT * FROM accounts WHERE phone = ?", (phone,))
    if account is None:
        return None
    return _sanitize_account(account)


async def update_account(
    db: DatabaseManager,
    account_id: str,
    data: Dict[str, Any],
) -> Optional[Dict[str, Any]]:
    """Update account fields.

    Only whitelisted fields are updated. Returns the updated account or None.
    """
    allowed_fields = {
        "display_name", "agent_name", "phone", "email", "public_key",
        "fingerprint", "status", "visit_permissions",
    }

    updates: list[str] = []
    params: list[Any] = []

    for field, value in data.items():
        if field not in allowed_fields:
            continue
        if field in ("visit_permissions",):
            value = json_serialize(value)
        updates.append(f"{field} = ?")
        params.append(value)

    if not updates:
        return await get_account(db, account_id)

    updates.append("updated_at = ?")
    params.append(iso_now())
    params.append(account_id)

    await db.execute(
        f"UPDATE accounts SET {', '.join(updates)} WHERE id = ?",
        params,
    )

    return await get_account(db, account_id)


async def delete_account(db: DatabaseManager, account_id: str) -> bool:
    """Delete an account and all related data.

    Returns True if the account was deleted, False otherwise.
    """
    # Delete sessions first (foreign key cascade should handle this, but be explicit)
    await db.execute("DELETE FROM sessions WHERE account_id = ?", (account_id,))
    # The accounts table has ON DELETE CASCADE on foreign keys referencing it,
    # but some tables (like node_permissions) don't have FK constraints.
    # Clean up related data explicitly.

    # Remove from friends arrays of other accounts
    account = await db.fetchone("SELECT friends FROM accounts WHERE id = ?", (account_id,))
    if account:
        friend_ids = json_list(account["friends"])
        for friend_id in friend_ids:
            await db.execute(
                "UPDATE accounts SET friends = ? WHERE id = ?",
                (json_serialize([]), friend_id),  # Simplified — ideally remove just this ID
            )

    # Remove node_permissions entries
    await db.execute("DELETE FROM node_permissions WHERE node_id = ? OR friend_id = ?", (account_id, account_id))

    # Remove friend_requests
    await db.execute("DELETE FROM friend_requests WHERE from_id = ? OR to_id = ?", (account_id, account_id))

    # Delete the account itself
    cursor = await db.execute("DELETE FROM accounts WHERE id = ?", (account_id,))
    deleted = cursor.rowcount > 0

    if deleted:
        logger.info("Deleted account: %s", account_id)

    return deleted


# ─── Internal Helpers ──────────────────────────────────────────────────


def _sanitize_account(account: Dict[str, Any]) -> Dict[str, Any]:
    """Remove sensitive fields from an account dict and deserialize JSON fields."""
    sanitized = dict(account)
    sanitized.pop("password_hash", None)

    # Deserialize JSON fields
    sanitized["friends"] = json_list(sanitized.get("friends"))
    sanitized["visit_permissions"] = json_list(sanitized.get("visit_permissions"))

    return sanitized
